File: server/app/controllers/posts_controller.py
from flask import jsonify
from app import db
from app.models import Post, User, Comment
from flask_jwt_extended import get_jwt_identity
from datetime import datetime
from bleach import clean
from app.utils.slugify import generate_unique_slug


def get_all_posts():
    posts = Post.query.filter_by(published=True).order_by(Post.created_at.desc()).all()

    if not posts or len(posts) == 0:
        return jsonify([]), 200

    result = []
    for post in posts:
        result.append(
            {
                "id": post.id,
                "title": post.title,
                "slug": post.slug,
                "content": post.content,
                "published": post.published,
                "author": post.user.username,
                "createdAt": post.created_at,
                "comments": [
                    {"id": c.id, "text": c.text, "createdAt": c.created_at}
                    for c in post.comments
                ],
            }
        )

    return jsonify(result), 200


def get_all_drafts():
    user_id = get_jwt_identity()
    drafts = (
        Post.query.filter_by(published=False, user_id=user_id)
        .order_by(Post.created_at.desc())
        .all()
    )

    if not drafts:
        return jsonify([]), 200

    return (
        jsonify(
            [
                {
                    "id": d.id,
                    "title": d.title,
                    "content": d.content,
                    "published": d.published,
                    "createdAt": d.created_at,
                }
                for d in drafts
            ]
        ),
        200,
    )


def get_post_by_slug(slug):
    post = Post.query.filter_by(slug=slug).first()
    if not post:
        return jsonify({"error": "Post not found."}), 404

    return (
        jsonify(
            {
                "id": post.id,
                "title": post.title,
                "content": post.content,
                "published": post.published,
                "author": post.user.username,
                "createdAt": post.created_at,
                "comments": [
                    {"id": c.id, "text": c.text, "createdAt": c.created_at}
                    for c in post.comments
                ],
            }
        ),
        200,
    )


def get_post_by_id(post_id):
    post = Post.query.get(post_id)
    if not post:
        return jsonify({"error": "Post not found."}), 404

    return (
        jsonify(
            {
                "id": post.id,
                "title": post.title,
                "content": post.content,
                "published": post.published,
                "author": post.user.username,
                "createdAt": post.created_at,
            }
        ),
        200,
    )


def create_post(data):
    user_id = get_jwt_identity()
    user = User.query.get(user_id)

    if not user or not user.is_author:
        return jsonify({"error": "User must be an author to perform this action."}), 403

    sanitized_content = clean(data.get("content", ""))

    # Generate a unique slug for the post
    title = data.get("title", "")
    slug = generate_unique_slug(title, db.session)

    post = Post(
        title=title,
        slug=slug,  # Set the slug here
        content=sanitized_content,
        published=bool(data.get("published", False)),
        user_id=user.id,
        updated_at=datetime.utcnow(),
    )
    db.session.add(post)
    db.session.commit()

    return jsonify({"message": "Post created."}), 200


def delete_post(post_id):
    user_id = get_jwt_identity()
    user = User.query.get(user_id)

    if not user or not user.is_author:
        return jsonify({"error": "User must be an author to perform this action."}), 403

    post = Post.query.get(post_id)
    if not post:
        return jsonify({"error": "Post not found."}), 404

    if post.user_id != user.id:
        return jsonify({"error": "You are not the author of this post."}), 403

    db.session.delete(post)
    db.session.commit()

    return jsonify({"message": "Post was successfully deleted."}), 200


def edit_post(post_id, data):
    user_id = get_jwt_identity()
    user = User.query.get(user_id)

    if not user or not user.is_author:
        return jsonify({"error": "User must be an author to perform this action."}), 403

    post = Post.query.get(post_id)
    if not post:
        return jsonify({"error": "Post not found."}), 404

    if post.user_id != user.id:
        return jsonify({"error": "You are not the author of this post."}), 403

    new_title = data.get("title", post.title)
    new_content = data.get("content", post.content)
    new_published = data.get("published", post.published)

    is_publishing = new_published is True and not post.published

    # If title changed, regenerate slug
    if new_title != post.title:
        post.title = new_title
        post.slug = generate_unique_slug(new_title, db.session)

    post.content = new_content
    post.published = new_published

    if is_publishing:
        post.created_at = datetime.utcnow()

    db.session.commit()

    return (
        jsonify(
            {
                "message": f"Post: {post.title} was successfully edited.",
                "post": {
                    "id": post.id,
                    "title": post.title,
                    "slug": post.slug,
                    "content": post.content,
                    "published": post.published,
                    "createdAt": post.created_at,
                },
            }
        ),
        200,
    )

File: server_node/controllers/postsController.js
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();
const asyncHandler = require("express-async-handler");
const CustomError = require("../utils/customError");
const { handleValidationErrors } = require("../utils/validator");
const { generateUniqueSlug } = require("../utils/slugify");
const DOMPurify = require("isomorphic-dompurify");
const { S3Client, DeleteObjectCommand } = require("@aws-sdk/client-s3");
const cheerio = require("cheerio");

const s3 = new S3Client({
  region: process.env.BUCKET_REGION,
  credentials: {
    accessKeyId: process.env.ACCESS_KEY,
    secretAccessKey: process.env.SECRET_ACCESS_KEY,
  },
});

const getAllPosts = asyncHandler(async (req, res) => {
  const posts = await prisma.post.findMany({
    where: { published: true },
    include: {
      comments: true,
      user: {
        select: { username: true },
      },
    },
    orderBy: { createdAt: "desc" },
  });

  posts.forEach((post) => {
    post.author = post.user.username;
    delete post.user;
  });

  if (!posts || posts.length === 0) {
    throw new CustomError(404, "No posts have been published.");
  }

  res.status(200).json(posts);
});

const getAllDrafts = asyncHandler(async (req, res) => {
  const drafts = await prisma.post.findMany({
    where: { published: false, userId: req.user.id },
    orderBy: { createdAt: "desc" },
  });

  if (!drafts || drafts.length === 0) {
    return;
  }

  res.status(200).json(drafts);
});

const getPostBySlug = asyncHandler(async (req, res) => {
  const slug = req.params.postSlug;
  const post = await prisma.post.findUnique({
    where: { slug },
    include: {
      comments: true,
      user: { select: { username: true } },
    },
  });

  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  post.author = post.user.username;
  delete post.user;

  res.status(200).json(post);
});

const getPostById = asyncHandler(async (req, res) => {
  const postId = parseInt(req.params.postId);
  const post = await prisma.post.findUnique({
    where: { id: postId },
    include: {
      user: {
        select: { username: true },
      },
    },
  });

  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  post.author = post.user.username;
  delete post.user;

  res.status(200).json(post);
});

const createPost = asyncHandler(async (req, res) => {
  handleValidationErrors(req);

  const user = await prisma.user.findUnique({ where: { id: req.user.id } });
  if (!user) {
    throw new CustomError(404, "User could not be found or does not exist.");
  }

  if (!user.isAuthor) {
    throw new CustomError(403, "User role must be Author to perform this action.");
  }

  const sanitizedContent = DOMPurify.sanitize(cleanHtmlContent(req.body.content));
  const slug = await generateUniqueSlug(req.body.title);
  const createdPost = await prisma.post.create({
    data: {
      title: req.body.title,
      slug,
      content: sanitizedContent,
      published: Boolean(req.body.published).valueOf(),
      userId: user.id,
      updatedAt: new Date(),
    },
  });

  res.status(200).json({ message: "Post created.", post: createdPost });
});

const deletePost = asyncHandler(async (req, res) => {
  const user = await prisma.user.findUnique({ where: { id: req.user.id } });
  if (!user.isAuthor) {
    throw new CustomError(403, "User role must be Author to perform this action.");
  }

  const post = await prisma.post.findUnique({
    where: { id: parseInt(req.params.postId) },
  });

  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  // Extract image URLs from post content, cheerio is used to parse html
  const $ = cheerio.load(post.content || "");
  const imageUrls = $("img")
    .map((_, el) => $(el).attr("src"))
    .get();

  // Convert URLs to S3 keys and delete
  const deletePromises = imageUrls.map((url) => {
    const key = new URL(url).pathname.slice(1); // remove leading '/'
    return s3.send(
      new DeleteObjectCommand({
        Bucket: process.env.BUCKET_NAME,
        Key: key,
      })
    );
  });

  try {
    await Promise.all(deletePromises);
  } catch (err) {
    console.error("Error deleting images from S3:", err);
  }

  // Delete post from database
  await prisma.post.delete({
    where: { id: parseInt(req.params.postId) },
  });

  res.status(200).json({ message: "Post was successfully deleted." });
});

const editPost = asyncHandler(async (req, res) => {
  handleValidationErrors(req);

  const user = await prisma.user.findUnique({ where: { id: req.user.id } });
  if (!user.isAuthor) {
    throw new CustomError(403, "User role must be Author to perform this action.");
  }

  const postId = parseInt(req.params.postId);
  const post = await prisma.post.findUnique({ where: { id: postId } });
  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  if (post.userId !== user.id) {
    throw new CustomError(403, "You are not the author of this post.");
  }

  const newTitle = req.body.title || post.title;
  let newSlug = post.slug;

  // Only generate a new slug if the title changed
  if (newTitle !== post.title) {
    newSlug = await generateUniqueSlug(newTitle, prisma);
  }

  const isPublishing = req.body.published === true && post.published === false;

  const updatedPost = await prisma.post.update({
    where: { id: postId },
    data: {
      title: newTitle,
      slug: newSlug,
      content: cleanHtmlContent(req.body.content || post.content),
      published: req.body.published === undefined ? post.published : Boolean(req.body.published),
      createdAt: isPublishing ? new Date() : post.createdAt,
    },
  });

  res.status(200).json({
    message: `Post: ${updatedPost.title} was successfully edited.`,
    post: updatedPost,
  });
});

const cleanHtmlContent = (html) => {
  const $ = cheerio.load(html);
  $("p").each(function () {
    if (!$(this).text().trim()) {
      $(this).remove();
    }
  });

  const cleaned = $.html().trim();
  return cleaned;
};

module.exports = {
  getAllPosts,
  getAllDrafts,
  getPostBySlug,
  getPostById,
  createPost,
  deletePost,
  editPost,
};
